### PR TITLE
Implement storage::ObjectWriteStream

### PR DESCRIPTION
**Pull request by coryan**
Introduce a IOStream writer (a class derived from `std::basic_ostream<char>`)
that uses libcurl pushing to GCS as its backend. This is part of the changes for
#557 to support streaming uploads.

### CODE REVIEW

**Comment by coryan** [comment]
@houglum This is ready for review now.

**Comment by coryan** [comment]
@houglum ping.

File: google/cloud/storage/internal/curl_streambuf.h [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+namespace storage {
+inline namespace STORAGE_CLIENT_NS {
+namespace internal {
+/**
+ * Implement a wrapper for libcurl-based streaming uploads.
+ */
+class CurlStreambuf : public ObjectWriteStreambuf {
+ public:
+  explicit CurlStreambuf(CurlUploadRequest&& upload,
+                         std::size_t max_buffer_size = 64 * 1024);
```
Comment (houglum) [inline google/cloud/storage/internal/curl_streambuf.h]: Any particular reason for 64K?

File: google/cloud/storage/internal/curl_streambuf.h [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+  bool IsOpen() const override;
+
+ protected:
+  int sync() override;
+  std::streamsize xsputn(char const* s, std::streamsize count) override;
+  int_type overflow(int_type ch) override;
+  HttpResponse DoClose() override;
+
+ private:
+  /// Raise an eception if the stream is closed.
```
Comment (houglum) [inline google/cloud/storage/internal/curl_streambuf.h]: s/eception/exception/

File: google/cloud/storage/internal/curl_upload_request.cc [hunk 3]
```diff
@@ -28,9 +28,15 @@ namespace internal {
 
 CurlUploadRequest::CurlUploadRequest()
     : headers_(nullptr, &curl_slist_free_all),
-      multi_(nullptr, &curl_multi_cleanup) {}
+      multi_(nullptr, &curl_multi_cleanup),
+      closing_(false),
+      curl_closed_(false) {
+  buffer_.reserve(128 * 1024);
```
Comment (houglum) [inline google/cloud/storage/internal/curl_upload_request.cc]: Why 128K?

File: google/cloud/storage/tests/run_integration_tests_production.sh [hunk 4]
```diff
@@ -23,6 +23,14 @@ echo
 echo "Running storage::internal::CurlRequest integration test."
 ./curl_request_integration_test
 
+#echo
```
Comment (houglum) [inline google/cloud/storage/tests/run_integration_tests_production.sh]: Add a TODO for what needs to be done before we can uncomment these?

Comment (coryan) [reply google/cloud/storage/tests/run_integration_tests_production.sh]: Actually I do not know if they ever should be running in the "production" integration tests. They would be hitting `https://nghttp2.org/httpbin` which seems wrong/abusive.  And we basically run the same thing in the `run_integration_tests_testbench.sh`.  I am inclined to remove all these curl tests from the `_production.sh` script, created #922 for that.

File: google/cloud/storage/internal/curl_upload_request.cc (hunk 3 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_upload_request.cc]: No particular reason, and also a bad a idea. Made it into a parameter configurable in the builder.

File: google/cloud/storage/internal/curl_streambuf.h (hunk 2 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_streambuf.h]: Fixed.

File: google/cloud/storage/internal/curl_streambuf.h (hunk 1 above)
Comment (coryan) [reply google/cloud/storage/internal/curl_streambuf.h]: Nope.  Bad idea actually, removed the default value.

**Pull request merged**